content/courses/ada-idioms/chapters/component_access_to_rec_objs.rst

```diff
@@ -0,0 +1,915 @@
+       begin
+          This.Transceiver.Transmit (HAL.UInt9 (Data));
+       end Put;
+
+       ---------
+       -- Get --
+       ---------
+
+       procedure Get (This : in out Device;  Data : out HAL.UInt8) is
+          Received : HAL.UInt9;
+       begin
+          This.Transceiver.Receive (Received);
+          Data := HAL.UInt8 (Received);
+       end Get;
+
+    end Serial_IO;
+
+
+Next, the interrupt-driven extension.
+
+.. code-block:: ada
+
+    with Ada.Interrupts;      use Ada.Interrupts;
+    with HAL;
+    with System; use System;
+
+    package Serial_IO.Interrupt_Driven is
+       pragma Elaborate_Body;
+
+       type Serial_Port
+         (Transceiver  : not null access USART;
+          IRQ          : Interrupt_ID;
+          IRQ_Priority : Interrupt_Priority)
+       is new Serial_IO.Device with private;
+       --  A serial port that uses interrupts for I/O. Extends the serial port
+       --  abstraction that is itself a wrapper for the USARTs hardware.
+
+       --  The procedures Initialize and Configure, among others, are implicitly
+       --  declared here, as operations inherited from Serial_IO.Device
+
+       overriding
+       procedure Put (This : in out Serial_Port;  Data : HAL.UInt8) with Inline;
+       --  Non-blocking, ie the caller can return before the Data goes out,
+       --  but does block until the underlying UART is not doing any other
+       --  transmitting. Does no polling. Will not interfere with any other I/O
+       --  on the same device.
+
+       overriding
+       procedure Get (This : in out Serial_Port;  Data : out HAL.UInt8) with Inline;
+       --  Blocks the caller until a character is available! Does no polling.
+       --  Will not interfere with any other I/O on the same device.
+
+    private
+
+       --  The protected type defining the interrupt-based I/O for sending and
+       --  receiving via the USART attached to the serial port designated by
+       --  Port. Each serial port object of the type defined by this package has
+       --  a component of this protected type.
+       protected type IO_Manager
+         (IRQ          : Interrupt_ID;
+          IRQ_Priority : Interrupt_Priority;
+          Port         : not null access Serial_Port)
+       --  with
+          --  Interrupt_Priority => IRQ_Priority   -- compiler bug :-(
+       is
+          pragma Interrupt_Priority (IRQ_Priority);
+
+          entry Put (Datum : HAL.UInt8);
+
+          entry Get (Datum : out HAL.UInt8);
+
+       private
+
+          Outgoing : HAL.UInt8;
+          Incoming : HAL.UInt8;
+
+          Incoming_Data_Available : Boolean := False;
+          Transmission_Pending    : Boolean := False;
+
+          procedure IRQ_Handler with Attach_Handler => IRQ;
+          --  The one interrupt handler for both sending and receiving. Calls the
+          --  respective Handle_* routines above for each incoming or outgoing
+          --  character interrupt on the USART.
```

Comment:
Let's delete all of this:  

"Calls the
          --  respective Handle_* routines above for each incoming or outgoing
          --  character interrupt on the USART."

because it is no longer true.

Reply:
Handled in #967.